cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1864,6 +1860,40 @@ Result<std::unique_ptr<KernelState>> MinMaxInit(KernelContext* ctx,
   return std::move(impl);
 }
 
+template <uint8_t index>
+Result<std::unique_ptr<KernelState>> MinOrMaxInit(KernelContext* ctx,
+                                                  const KernelInitArgs& args) {
+  std::vector<ValueDescr> inputs = args.inputs;
+
+  ARROW_ASSIGN_OR_RAISE(auto func, GetFunctionRegistry()->GetFunction("hash_min_max"));
+  ARROW_ASSIGN_OR_RAISE(auto kernel, func->DispatchBest(&inputs));
+
+  KernelInitArgs new_args{kernel, inputs, args.options};
+  return kernel->init(ctx, new_args);
+}
+
+template <uint8_t index>
+HashAggregateKernel MakeMinOrMaxKernel() {
+  HashAggregateKernel kernel;
+  kernel.init = MinOrMaxInit<index>;
+  kernel.signature = KernelSignature::Make(
+      {InputType(ValueDescr::ANY), InputType::Array(Type::UINT32)},
+      OutputType([](KernelContext* ctx,
+                    const std::vector<ValueDescr>& descrs) -> Result<ValueDescr> {
+        return ValueDescr::Array(descrs[0].type);
+      }));
```

Comment:
Nit: Here `OutputType` is expressed as a lambda function, so should [`MinMaxType`](https://github.com/apache/arrow/pull/11152/files#diff-fead2cb629ddc5a706a9383d4e64e24064cb3a2103e5e078bde7a11b57fd9654R297) be also a lambda.